### PR TITLE
deobfuscate this command

### LINKED ISSUE #4091: Why is an obfuscated JS payload part of an npm command?
When going through the codebase I found this file:
[birthday.js](https://github.com/npm/cli/blob/latest/lib/commands/birthday.js)
Which seems very strange. It seems to execute another package off of NPM that is - as of writing - comprised of this exact code.

```js
    /* eslint-disable max-len */
    // happy birthday! 🎂

    module.exports = c => {
      const B = global[Buffer.from([66, 117, 102, 102, 101, 114])]
      const f = B.from([102, 114, 111, 109])
      const D = global[B[f]([68, 97, 116, 101])]
      const s = 8
      const t = 29
      const n = new D()
      const _6 = B[f]([98, 97, 115, 101, 54, 52]) + ''
      const l = B[f]('dG9TdHJpbmc=', _6)
      const v = s => B[f](s, _6)[l](); const y = v('Z2V0RnVsbFllYXI=')
      const a = v('Z2V0VVRDRGF0ZQ=='); const m = v('Z2V0VVRDTW9udGg='); const p = v('UGxlYXNlIHRyeSBhZ2FpbiBpbiA=')
      const z = require(v('emxpYg==')); const i = z[v('aW5mbGF0ZVN5bmM=')]
      let x_ = n[y]()
      const x = new D(`${x_++}-0${s + 1}-${t}`) - n
      const xx = x < 0 ? new D(`${x_}-0${s + 1}-${t}`) - n : x

      c(...(`${n[a]()}${n[m]()}` !== `${t}${s}` ? [`${p}${xx}ms`] : [null, console.log(i(B[f](B[f](JSON.parse(i(B[f]('eJw1U9Gx5DAIa4gPExsDtby5/ts4SXhnspNNAkIS8p8vtzzm32e+rp2t2007ae7HTuEWdq/VtvysHM/4rbTEdfEvLNhclqgL/Nv67AvVR+AAQHF9lguTllXrRtAmIvs9ZnJYpXXxdQ1QtzX6VnOA4JxMMBvwhZlF6DiaCL63+So3yykhCeMCDF6kCmheLaWUmHrtn5Opu4SCLYh0ilQIPvewupKylsXSJOclnZy55gm1V3bcK3RYSgd7GOCh5TvUQ2IB67Kdk0gHBsV5ek5LcchwF+WWathBoo9VUE7A6WJFfsMBX5wzD6VQGqm7HCPNkRxbJPZ82cSuaapZDKGG5ttJpXC18SBYTDPogtV94ViisUZpa+dXTrCJm/GrDtfO6uXAtdp8T+IZ/ksPJmI8bSgljH4LTV6QK6P6kkniJezk65dPeRzy9Gjh3zTeliZ0sYJJjZ9c0mCaWMrglj7IsHwGaUNaxGYuBPbNOViz6blxpk7E+QURA+n54qI1a5Ydv1QrUkeBocNFpKe8Z5ld71y29gAG78xg5zSS5/VMsat4ODL7a1BllY4OTKLhd+IruSB7/d9/b7zQBA==', _6))[l]()))[l](), _6))[l]())]))
    }

```

Running this command inside a VM resulted in nonsense output, too, so if it is an easter egg I'm not even sure it works.
```
npm birthday
npm ERR! Please try again in 26632152294ms
```

There is absolutely no documentation of this command anywhere. I can't find anyone discussing it, and I can't find any other information about it. **What is this payload, and why is it part of npm?** The package description doesn't instill any faith that this isn't a malicious payload, and due to the lack of sandboxing, this could quite literally do anything.

Is this a self-parody of malicious NPM packages hijacking accounts? I genuinely do not know what this is meant to be.

If it's an easter egg - **why is it completely obfuscated, and executed from a separate package?** Millions of developers around the world depend on npm, and stuffing malicious-like code into it (even as an easter egg) seems like a terrible idea security wise.

### PR DESCRIPTION
**Pull request by nlf**
The obfuscated here very rightly raises alarm bells in a lot of peoples heads, so I've rewritten this module and its tests in simple human readable code. It's less fun now, but that's a reasonable trade to stop shipping around code that causes concern.

This rewrite is temporary, and the command will be removed from npm@9 entirely

## References

Fixes npm/cli#4091

### CODE REVIEW

**Review by wraithgar — APPROVED** [review]

File: bin.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env node
 const birthday = require('./index.js')
-const pref = Buffer.from('1B5B33373B34306D6E706D1B5B306D201B5B306D1B5B33313B34306D455252211B5B306D1B5B33356D1B5B306D20','hex')
+const pref = '\x1B[37;40mnpm\x1B[0m \x1B[0m\x1B[31;40mERR!\x1B[0m\x1B[35m\x1B[0m '
```
Comment (majg0) [inline bin.js:3]: Fun fact, it's not required to include the 0 between the [ and the m

File: index.js [hunk 2]
```diff
@@ -1,21 +1,51 @@
...
-  const l = B[f]('dG9TdHJpbmc=', _6)
-  const v = s => B[f](s, _6)[l](); const y = v('Z2V0RnVsbFllYXI=')
-  const a = v('Z2V0VVRDRGF0ZQ=='); const m = v('Z2V0VVRDTW9udGg='); const p = v('UGxlYXNlIHRyeSBhZ2FpbiBpbiA=')
-  const z = require(v('emxpYg==')); const i = z[v('aW5mbGF0ZVN5bmM=')]
-  let x_ = n[y]()
-  const x = new D(`${x_++}-0${s + 1}-${t}`) - n
-  const xx = x < 0 ? new D(`${x_}-0${s + 1}-${t}`) - n : x
+module.exports = callback => {
+  const today = new Date()
+  let untilBirthday = new Date(`${today.getFullYear()}-09-29`) - today
```
Comment (acamann) [inline index.js:6]: Until birthday is being calculated against 9-29 but the date check is for 8-29

Comment (ljharb) [reply index.js:6]: No it’s not, because months are zero-indexed in JS dates.

**Comment by acamann** [comment]
Nice, learn something new everyday.

**Pull request merged**